### PR TITLE
Add deferrable mode to `BatchSensor` 

### PR DESCRIPTION
**Pull request by phanikumv**
This PR donates the `BatchAsyncSensor`  from [astronomer-providers](https://github.com/astronomer/astronomer-providers) to Airflow

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/amazon/aws/triggers/batch.py [hunk 1]
```diff
@@ -105,3 +105,92 @@ async def run(self):
...
+        async with self.hook.async_conn as client:
+            waiter = self.hook.get_waiter("batch_job_complete", deferrable=True, client=client)
+            attempt = 0
+            while attempt < self.max_retries:
+                attempt = attempt + 1
+                try:
+                    await waiter.wait(
+                        jobs=[self.job_id],
+                        WaiterConfig={
+                            "Delay": self.poke_interval,
```
Comment (pankajastro) [inline airflow/providers/amazon/aws/triggers/batch.py]: I'm not sure if it accepts float or not. The delay should be int

File: airflow/providers/amazon/aws/sensors/batch.py [hunk 2]
```diff
@@ -75,6 +84,32 @@ def poke(self, context: Context) -> bool:
...
+                method_name="execute_complete",
+            )
+
+    def execute_complete(self, context: Context, event: dict[str, Any]) -> None:
+        """
+        Callback for when the trigger fires - returns immediately.
+        Relies on trigger to throw an exception, otherwise it assumes execution was
+        successful.
+        """
+        if "status" in event and event["status"] == "error":
```
Comment (pankajastro) [inline airflow/providers/amazon/aws/sensors/batch.py]:
```suggestion
        if "status" in event and event["status"] == "failure":
```
no?

File: airflow/providers/amazon/aws/hooks/batch_client.py [hunk 3]
```diff
@@ -577,3 +579,77 @@ def exp(tries):
         delay = 1 + pow(tries * 0.6, 2)
         delay = min(max_interval, delay)
         return uniform(delay / 3, delay)
+
+
+class BatchClientAsyncHook(BatchClientHook, AwsBaseAsyncHook):
```
Comment (pankajastro) [inline airflow/providers/amazon/aws/hooks/batch_client.py]: where we are using this?

Comment (phanikumv) [reply airflow/providers/amazon/aws/hooks/batch_client.py]: removed it!

File: airflow/providers/amazon/aws/triggers/batch.py (hunk 1 above)
Comment (phanikumv) [reply airflow/providers/amazon/aws/triggers/batch.py]: Good catch ! this change is done.

File: airflow/providers/amazon/aws/sensors/batch.py (hunk 2 above)
Comment (phanikumv) [reply airflow/providers/amazon/aws/sensors/batch.py]: Fixed

**Comment by potiuk** [comment]
Tests/static checks are failing.

**Comment by phanikumv** [comment]
> Tests/static checks are failing.

yes I am checking it

**Comment by phanikumv** [comment]
> Tests/static checks are failing.

this is fixed

File: airflow/providers/amazon/aws/sensors/batch.py [hunk 4]
```diff
@@ -41,6 +43,7 @@ class BatchSensor(BaseSensorOperator):
     :param job_id: Batch job_id to check the state for
     :param aws_conn_id: aws connection to use, defaults to 'aws_default'
     :param region_name: aws region name associated with the client
+    :param deferrable: Run sensor in the deferrable mode.
```
Comment (jedcunningham) [inline airflow/providers/amazon/aws/sensors/batch.py:46]: Why aren't `poke_interval` and `max_retries` here?

File: airflow/providers/amazon/aws/triggers/batch.py [hunk 5]
```diff
@@ -105,3 +105,92 @@ async def run(self):
...
             yield TriggerEvent({"status": "success", "job_id": self.job_id})
+
+
+class BatchSensorTrigger(BaseTrigger):
+    """
+    Checks for the status of a submitted job_id to AWS Batch until it reaches a failure or a success state.
+    BatchSensorTrigger is fired as deferred class with params to poll the job state in Triggerer.
+
+    :param job_id: the job ID, to poll for job completion or not
+    :param aws_conn_id: connection id of AWS credentials / region name. If None,
```
Comment (jedcunningham) [inline airflow/providers/amazon/aws/triggers/batch.py]: These are out of order.

File: airflow/providers/amazon/aws/triggers/batch.py
Comment (jedcunningham) [inline airflow/providers/amazon/aws/triggers/batch.py:1]: I don't see any test coverage for this trigger?

Comment (phanikumv) [reply airflow/providers/amazon/aws/triggers/batch.py:1]: Fixed in [388cbaa](https://github.com/apache/airflow/pull/30279/commits/388cbaa3572fd7e46985e33edadd92ec372d7169)

File: airflow/providers/amazon/aws/triggers/batch.py (hunk 5 above)
Comment (phanikumv) [reply airflow/providers/amazon/aws/triggers/batch.py]: Fixed in [388cbaa](https://github.com/apache/airflow/pull/30279/commits/388cbaa3572fd7e46985e33edadd92ec372d7169)

File: airflow/providers/amazon/aws/sensors/batch.py (hunk 4 above)
Comment (phanikumv) [reply airflow/providers/amazon/aws/sensors/batch.py:46]: Fixed in [388cbaa](https://github.com/apache/airflow/pull/30279/commits/388cbaa3572fd7e46985e33edadd92ec372d7169)

File: airflow/providers/amazon/aws/triggers/batch.py [hunk 6]
```diff
@@ -105,3 +105,92 @@ async def run(self):
...
+        """
+        Make async connection using aiobotocore library to AWS Batch,
+        periodically poll for the Batch job status.
+
+        The status that indicates job completion are: 'SUCCEEDED'|'FAILED'.
+        """
+        async with self.hook.async_conn as client:
+            waiter = self.hook.get_waiter("batch_job_complete", deferrable=True, client=client)
+            attempt = 0
+            while attempt < self.max_retries:
```
Comment (pankajastro) [inline airflow/providers/amazon/aws/triggers/batch.py]: just posting some discussion here https://github.com/apache/airflow/pull/30945#discussion_r1218325194. What would happen if the trigger stops and picked up again on another machine then this max_retries would not make much sense. wdyt?

File: airflow/providers/amazon/aws/sensors/batch.py [hunk 7]
```diff
@@ -75,6 +87,32 @@ def poke(self, context: Context) -> bool:
 
         raise AirflowException(f"Batch sensor failed. Unknown AWS Batch job status: {state}")
 
+    def execute(self, context: Context) -> None:
+        if not self.deferrable:
+            super().execute(context=context)
+        else:
+            self.defer(
+                timeout=timedelta(seconds=self.timeout),
```
Comment (pankajastro) [inline airflow/providers/amazon/aws/sensors/batch.py]: I think we should calculate this timeout based on  poke_interval and max_retries if given, wdyt?

Comment (pankajastro) [reply airflow/providers/amazon/aws/sensors/batch.py]: similar feedback are here https://github.com/apache/airflow/pull/30945#discussion_r1218436958

File: airflow/providers/amazon/aws/triggers/batch.py (hunk 6 above)
Comment (phanikumv) [reply airflow/providers/amazon/aws/triggers/batch.py]: done in 
9dc72f5

File: airflow/providers/amazon/aws/sensors/batch.py (hunk 7 above)
Comment (phanikumv) [reply airflow/providers/amazon/aws/sensors/batch.py]: done in 
9dc72f5

**Review by pankajastro — APPROVED** [review]
LGTM

**Pull request merged**